### PR TITLE
RFC 24 Prototype

### PR DESCRIPTION
**Pull request by shaobo-he-aws**
## Description of changes
This PR implements a parser and a pretty printer of the custom schema syntax proposed in https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md.

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by andrewmwells-amazon — APPROVED** [review]

**Review by aaronjeline — APPROVED** [review]

**Comment by john-h-kastner-aws** [comment]
Does this parser provide access to source ranges in a way that they could be stored on the parsed schema and used for error reporting later? This is lacking from the JSON schema format, and we have said before that we want to include this type of error reporting when we introduce the custom format.

**Comment by shaobo-he-aws** [comment]
> Does this parser provide access to source ranges in a way that they could be stored on the parsed schema and used for error reporting later? This is lacking from the JSON schema format, and we have said before that we want to include this type of error reporting when we introduce the custom format.

It doesn't do it now but shouldn't be hard to implement. I'll make sure we have this feature for the final version.

File: cedar-policy-validator/src/schema_syntax.rs [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+            optional(choice((
+                (accept(Token::Comma), AppParser()).map(|(_, cdr)| cdr),
+                accept(Token::Comma).map(|_| ApplySpec {
+                    resource_types: None,
+                    principal_types: None,
+                    context: AttributesOrContext::default(),
+                }),
+            ))),
+        )
+            .map(|(singleton, cdr)| {
```
Comment (john-h-kastner-aws) [inline cedar-policy-validator/src/schema_syntax.rs:323]: This logic is a bit confusing.

**Review by john-h-kastner-aws — COMMENTED** [review]
Looks nice. The approach or parsing into the `SchemaFragment` and `NamespaceDefinition` is good. This should let us share the rest of the schema construction code.

The parser combinators library is nice, but I do think it's better to use lalrpop for the policy and schema grammar like we said earlier. Not sure what the performance implication are, but i wouldn't want to end up using two parser libraries.

**Pull request closed without merging**